### PR TITLE
Fix automation for nofile/matchid

### PR DESCRIPTION
**Pull request by flodolo**
Unfortunately, it turns out I was naive in my rewrite, completely breaking the `nofile`/`matchid` options, and introducing some unexpected behavior even in standard mode.

## `standard` mode

When a trans-unit moves between `<file>` elements, the translation gets removed. This is unnecessary, since the goal is to minimize the diff, and let Pontoon deal with clean-ups.

## `nofile`/`matchid` mode

In the same scenario, the translation is not carried over, since we're not rebuilding the localized file from reference anymore 🤦‍♂️.

This rewrites fixes both issues. In order to make portability between iOS and VPN easier, I've centralized some config (locale mapping, excluded folders).

I've also created a separate repository with tests, given how many weird edge cases I hit in the process: https://github.com/mozilla-l10n/xliff-l10n-scripts

### CODE REVIEW

File: .github/scripts/update_other_locales.py
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py]: Now that I know this code better I am suggesting some general changes for clarity (some of these are on lines you did not touch, sorry about that)

```suggestion
        help="Locale code for source strings (usually en-US)",
```

Comment (camilapedraza) [inline .github/scripts/update_other_locales.py]:
```suggestion
        help="Path to folder containing subfolders for all locales",
```

File: .github/scripts/update_other_locales.py [hunk 1]
```diff
@@ -62,17 +321,29 @@ def main():
...
+    parser.add_argument(
+        "--project",
+        required=False,
+        default=None,
+        choices=sorted(PROJECTS),
+        help="Project config to use (locale mapping + excluded folders).\n"
+        "Defaults to no mapping and no excluded folders.",
     )
 
     parser.add_argument("locales", nargs="*", help="Locales to process")
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py]:
```suggestion
    parser.add_argument("locales", nargs="*", help="Locales to process - if none listed, all locale subfolders in the path will be processed")
```

File: .github/scripts/update_other_locales.py [hunk 2]
```diff
@@ -5,44 +5,303 @@
...
-   - 'standard': keep the translation only if a reference entry matches both
-     file and source. Removes it if the source changed or the string moved to
-     a different file.
-   - 'nofile': keep the translation if a reference entry matches the source,
-     ignoring the file. This retains translations when a string moves as-is
-     from one file to another.
-   - 'matchid': always keep the translation, and update <source> to the
-     reference text if it changed. This retains translations for trivial
-     source changes.
+update_other_locales.py --reference <locale> --path <folder> [locales...]
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py]: Minor/non-blocking: the module docstring synopsis (and body) doesn't mention --project (or --type)

```suggestion
update_other_locales.py --reference <locale> --path <folder>
      [--type standard|nofile|matchid] [--project <name>] [locales...]
```

File: .github/scripts/update_other_locales.py [hunk 3]
```diff
@@ -5,44 +5,303 @@
...
+
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:14]:
```suggestion

--project selects the locale mapping and excluded folders from locale_config.py. If no project name is provided, empty defaults will be returned.

```

File: .github/scripts/update_other_locales.py [hunk 4]
```diff
@@ -113,101 +376,56 @@ def main():
...
-            # Replace the existing locale file with the updated XML content
-            write_xliff(locale_tree, l10n_file)
-            updated_files += 1
+                # Resolve the folder name to its XLIFF target-language code.
+                locale_code = get_locale_code(mapping, locale)
+
+                new_tree = rebuild_from_reference(
+                    reference_tree, locale_root, update_type, locale_code
+                )
+                write_xliff(new_tree, l10n_file)
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:488]: I am sure there is a reason rebuild_from_reference deepcopies the whole ElementTree rather than the root element, but I can't pintpoint it. Since write_xliff re-adds the declaration manually, I would think copying the root would suffice but there must be stuff I am not thinking of. 🤔

File: .github/scripts/update_other_locales.py [hunk 5]
```diff
@@ -5,44 +5,303 @@
...
+            )
+            continue
+
+        reference_sources = per_file.get(file_original(trans_node))
+        if reference_sources is None:
+            # The ID exists in the reference but only in a different <file>: the
+            # string moved, leave it in place ('nofile'/'matchid' relocate it).
+            continue
+
+        # Same file: remove only when the source text actually changed here.
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:173]: From the logic flow here I understand that, in standard mode, a string that's both _moved to another file_ **and** _had its source text edited_ keeps its old translation. Is that intended?

Example scenario 1: "welcome_text": "Wellcome" in file A gets moved to file B with the same id but the typo in "Wellcome" is fixed.
Example scenario 2: "welcome_text" "Hi!" in file A gets moved to file B with the same id but the text significantly changes to "Welcome to Firefox!".

This check in both cases will leave the translations untouched for standard mode because it checks only whether the id moved files (and skips the string if a move happened) before checking if the text changed. I know we have nofile and matchid for scenario 1, but what about a genuine lapse of judgement like scenario 2 that goes by undetected by the standard mode? Is that intended?

File: .github/scripts/update_other_locales.py [hunk 6]
```diff
@@ -5,44 +5,303 @@
...
+    changed in the reference for the same XLIFF file. Strings removed upstream
+    or moved to a different file are left untouched.
+    """
+    for trans_node in locale_root.xpath("//x:trans-unit", namespaces=NS):
+        target = trans_node.find("x:target", namespaces=NS)
+        if target is None:
+            # Untranslated string, nothing to do.
+            continue
+
+        per_file = reference_index.get(trans_node.get("id"))
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py]: A suggestion for readability on variable names:

Changing `per_file` here and in `update_in_file` to something like `sources_by_file`. I kept having to retrace the origin of this variable to remember what was stored in it (but it might just be me). This is minor/non-blocking!

File: .github/scripts/update_other_locales.py (hunk 5 above)
Comment (camilapedraza) [reply .github/scripts/update_other_locales.py:173]: On a related note: I know we want small diffs so we leave any restructuring to the matchid and nofile modes. But since we are mainly running everything through standard (and these scripts are supposed to catch and fix stuff that humans might have missed) are there any concerns around strings living in two different "files" across locales long term?

For example, if scenario 2 above happens in two different moments: first we move the string "welcome_text" from file A to B without text changes and use standard mode (nothing happens, translations are preserved). Days, weeks, months later that string's text is changed from "Hi!" to "Welcome to Firefox" (still in file B and no ID changes). From what I understand standard mode won't detect this (translations will be unaligned with source).

Also this highlights some gaps in my understanding of what Pontoon does and doesn't do (and which maybe are not fully clarified in the docstring above):
Will Pontoon avoid this de-sync between translation and source, updating file from A to B in the localized files? Or do we need to rely on nofile mode for that and the degree of divergence continues to increase until we do?

File: .github/scripts/update_other_locales.py [hunk 7]
```diff
@@ -113,101 +376,56 @@ def main():
...
-                (file_name, source.text)
-            )
+        # 'standard' only needs an index of the reference sources per ID.
+        reference_index = (
+            build_reference_index(reference_root, filename)
+            if update_type == "standard"
+            else None
+        )
 
         for locale in locales:
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:458]: Minor/non-blocking: A lot of the code is repeated for both branches of this if/else logic flow. Maybe we can refactor for readability? a suggestion:

```
  for locale in locales:
      l10n_file = os.path.join(base_folder, locale, filename)

      if not os.path.isfile(l10n_file):
          continue

      try:
          locale_tree = etree.parse(l10n_file)
          locale_root = locale_tree.getroot()
      except Exception as e:
          print(f"ERROR: Can't parse {l10n_file}")
          print(e)
          continue

      if update_type == "standard":
          # In-place update, requires an existing localized file.
          print(f"Processing {l10n_file}")
          update_in_place(reference_index, locale_root)
          write_xliff(locale_tree, l10n_file)
      else:
          # Rebuild from reference, moving existing translations.
          # Requires an existing localized file
          print(f"Updating {l10n_file}")

          # Resolve the folder name to its XLIFF target-language code.
          locale_code = get_locale_code(mapping, locale)
          new_tree = rebuild_from_reference(
              reference_tree, locale_root, update_type, locale_code
          )
          write_xliff(new_tree, l10n_file)

      updated_files += 1
```

we can also move the print message outside of the conditional logic if we are okay with keeping unified wording ("Processing {l10n_file}")

File: .github/scripts/update_other_locales.py [hunk 8]
```diff
@@ -113,101 +376,56 @@ def main():
...
+                # Resolve the folder name to its XLIFF target-language code.
+                locale_code = get_locale_code(mapping, locale)
+
+                new_tree = rebuild_from_reference(
+                    reference_tree, locale_root, update_type, locale_code
+                )
+                write_xliff(new_tree, l10n_file)
+                updated_files += 1
 
     if updated_files == 0:
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:492]: This is pre-existing code but your changes to make this script reusable for other projects (including brand new projects) means the first import for a never-localized project would exit non-zero and fail CI. 

I am not familiar with how we set up new projects for localization so maybe there are steps in between setup and this code running that would prevent this. If not,
should we address this?

File: .github/scripts/update_other_locales.py [hunk 9]
```diff
@@ -62,17 +321,29 @@ def main():
         required=False,
         default="standard",
         dest="update_type",
-        help="""Type of update. Existing translation is maintained if:
-    - 'standard': matches file, ID, and source text
-    - 'nofile': matches ID and source text, ignoring file
-    - 'matchid': matches ID""",
+        help="""Type of update:
+    - 'standard': in place, remove a translation only if the source changed
+    - 'nofile': rebuild from reference, move translations if ID and source text match
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:392]: Probably insane of me to be suggesting a change like this at this stage but I find "nofile" very confusing. Among other things because both matchid and nofile ignore the file. 

I would suggest maybe using something like matchsource (to align with matchid and be more descriptive about what it does). But obviously this might change things beyond just the name (how people are currently using it, any automations in place, etc.) so I know this is risky and probably unnecessary.

**Review by camilapedraza — CHANGES_REQUESTED** [review]
First pass here focusing on standard mode only. Before I move into a deep dive for nofile and matchid, I would love to look at all this stuff and maybe float the idea that nofile is living in a weird zombie mode with your latest changes, and maybe needs to be rethought entirely:

(This might be a controversial and unpopular opinion, so bear with me)
if we WANT to preserve translations when a string moves files and the text is unchanged, and if the current way standard behaves creates some potential issues with moved strings (see in-line comments), shouldn't we just move unchaged strings to their new files by default and avoid divergence between source and translated files altogether?

I know this goes AGAINST the small diffs proposal but I don't think diffs with strings changing files are too hard to review, and wanting to avoid a large diff is potentially creating more problems in the long run.

File: .github/scripts/update_other_locales.py (hunk 5 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:173]: > From the logic flow here I understand that, in standard mode, a string that's both moved to another file and had its source text edited keeps its old translation. Is that intended?

No, that's a bug that needs to be fixed. It's not a real case scenario we have seen before, but it should be prevented. Great catch!

> For example, if scenario 2 above happens in two different moments: first we move the string "welcome_text" from file A to B without text changes and use standard mode (nothing happens, translations are preserved). Days, weeks, months later that string's text is changed from "Hi!" to "Welcome to Firefox" (still in file B and no ID changes). From what I understand standard mode won't detect this (translations will be unaligned with source).

Even with the current buggy code, in the second part:
- Both `standard` and `nofile` would remove the translation (`<target>`) from the localized XLIFF files, since the source text changed. 
- Pontoon will show the existing translation as rejected after the next sync.

File: .github/scripts/update_other_locales.py (hunk 4 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:488]: Copying the root should work in the same. `write_xliff` only calls `etree.indent()` and `etree.tostring()`, both of which accept an ElementTree or an Element. 

The one hard requirement is not mutating `reference_tree`, which is parsed once per file and reused across every locale. So it's purely stylistic (I keep it a tree just to avoid a .getroot() round-trip).

File: .github/scripts/update_other_locales.py (hunk 8 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:492]: The last time we created a new XLIFF project was 6 years ago with VPN (this repo is 11 years old). I guess we never really thought about that scenario, but it's worth fixing 😅

File: .github/scripts/update_other_locales.py (hunk 6 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py]: I agree that `per_file` is not a good name, but I have a feeling `sources_by_file` might be misleading, since the dictionary is organized by ID, the file, e.g. 

```
{
    'App Icon': {
        'Client/en.lproj/InfoPlist.strings': {
            'App Icon'
        }
    }, 
    'NSCameraUsageDescription': {
        'Client/en.lproj/InfoPlist.strings': {
            'Firefox can use your camera to scan QR codes or take photos and video.'
        }
    }, 
    'CFBundleDisplayName': {
        'Extensions/ActionExtension/en.lproj/InfoPlist.strings': {
            'Open In Firefox'
        }
    }
}
```

I'm going with `sources_by_id` for now.

File: .github/scripts/update_other_locales.py (hunk 9 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:392]: We can definitely change it, since we own all the repositories using this automation.

With that said, I don't `matchsource` is clearer. `nofile` is supposed to carry over the translation ignoring the `file`.

**Comment by flodolo** [comment]
I've tried to address most of comments before we meet. Also renamed some of the variables in the rest of the code, hoping making them clearer.

`carry_over_obsolete()` is the overkill part of the code, but we can discuss it (it has benefits, and it works). In the meantime, I'm keeping changes (+tests) up to date in https://github.com/mozilla-l10n/xliff-l10n-scripts/pull/1/changes

As a next step, it would be nice to actually centralize the code, either cloning that repo as part of automation (easy path), or make it a PyPI package (possibly overkill).

File: .github/scripts/update_other_locales.py [hunk 10]
```diff
@@ -5,74 +5,379 @@
...
+    for strings removed upstream are carried over (see carry_over_obsolete).
+    That prevents automation from touching localized files, leaving the removal
+    to Pontoon instead, and reducing merge conflicts.
+
+    'locale_root' is the current localized content of an existing file.
+    """
+    # Remember each localized <file>'s attribute order, to restore it on the
+    # rebuilt tree (which otherwise inherits the reference's order).
+    # Without this, different automations (this script, extraction, Pontoon)
+    # would start fighting over the attribute order, creating unnecessary diffs.
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:283]: I understand why this is necessary, but after reviewing the PR I started asking myself why we find ourselves in this situation to begin with.

Is the difference in attribute order being created by Pontoon? If so, why is Pontoon structuring target locale files with the attributes in a different order than that of the source file? If not, does this divergence come from our own scripts when importing? 
Would it be worth:
- batch-normalizing target files to match source files, and
- ensuring Pontoon/any import scripts respect source attribute order in target files

so that we are not tasked with preserving arbitrary attribute in scripts like this one?

File: .github/scripts/update_other_locales.py [hunk 11]
```diff
@@ -5,74 +5,379 @@
...
+
+        # Same file: remove only when the source text actually changed here.
+        if source_node.text not in files_for_id:
+            target.getparent().remove(target)
+
+
+def carry_over_obsolete(new_root, locale_root, reference_ids, locale_code):
+    """
+    Keep strings that no longer exist in the reference (removed upstream) in the
+    rebuilt localized file, so their removal is left to Pontoon instead of the
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:181]: You've probably explained this a thousand times already but double checking the effects of this:
Does this mean we're avoiding a large diff in this phase of the process only to have a large diff when Pontoon performs the removal and pushes the updated locale files back to the repo?

File: .github/scripts/update_other_locales.py [hunk 12]
```diff
@@ -5,74 +5,379 @@
...
+            if existing_target is not None:
+                existing_target.text = translation
+            else:
+                # Insert a new <target> right after <source>, in the XLIFF
+                # namespace so the in-memory tree matches what is written to
+                # disk (a bare etree.Element("target") would be namespaceless
+                # in memory and only pick up the default namespace on save).
+                target = etree.Element(f"{{{NS['x']}}}target")
+                target.text = translation
+                source_node.addnext(target)
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:350]: You added a guard in line 287 in case source_node is None, but here we are calling addnext on something that can be None, which will throw an attribute error and the run will crash.

I believe this kind of malformed string is highly unlikely, but we did guard against this in the "standard path" (line 148 in update_in_place() function). Do we need guard here as well?

File: .github/scripts/update_other_locales.py [hunk 13]
```diff
@@ -5,74 +5,379 @@
...
+        if new_dest is None:
+            # The whole <file> block was removed from the reference: recreate
+            # it and empty its <body> (no <trans-unit> elements), preserving
+            # the <file> attributes. Obsolete strings will be reinserted later.
+            new_dest = deepcopy(loc_file)
+            new_dest.set("target-language", locale_code)
+            new_dest_body = new_dest.find("x:body", namespaces=NS)
+            for tu in new_dest_body.xpath("./x:trans-unit", namespaces=NS):
+                new_dest_body.remove(tu)
+            new_root.append(new_dest)
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py]: Want to confirm that there will be cases in which:
- we have a <file> node that was deleted in the source, and (because all of it's strings were not moved but truly deleted they are now obsolete) therefore
- that deleted <file> it will be recreated by this function, and
- the recreated <file> node will appended at the END of the new_root

This means the file node will exist for Pontoon to handle but it will also still show up in the diff for all localized files in Github (as a deletion from its original place and an insertion at the bottom of the file)

Am I right? Does this go against the minimal diff objective? or does this still hold water regardless?

File: .github/scripts/update_other_locales.py [hunk 14]
```diff
@@ -5,74 +5,379 @@
...
+    string is matched wherever it lives now.
+    """
+    if update_type == "matchid":
+        # Ignore source text: retain the translation even if the source changed.
+        return original_id
+    # 'nofile': invalidate on source change.
+    return f"{original_id}:{hash(source_string)}"
+
+
+def file_original(trans_node):
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py]: Claude made me realize that the way we structured this, we are walking down thousands of trans_nodes and finding their original files (times however many hundreds of locales we have). 

Not sure it's worth refactoring right now, but would it maybe make sense to leave a note for future refactors to maybe iterate by file, read original once for the file, and then loop over all the units of the file to store the read original for each? Not sure we even gain much performance by doing that but wanted to flag it since I know we've discussed some of these runs taking a significant amount of time and I don't know how much things like this factor into that time.

**Review by camilapedraza — COMMENTED** [review]
No actual changes required, just a bunch of questions/doubts/sanity-checks. We can go over this during our 1-1 if needed!
Hope I am not creating confusion or asking you to reiterate the logic for the n-th time. This is still a lot of new context for my newbie brain 🤯

File: .github/scripts/update_other_locales.py (hunk 11 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:181]: Yes.

The PR will only show strings changed or added. Once the PR is merged, Pontoon will go and clean up all localized files, putting them in sync with English, and generating a much larger diff (that we don't need to review).

File: .github/scripts/update_other_locales.py (hunk 13 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py]: You're absolutely correct. This is appending it at the end, so it still creates an unnecessary diff. 

I *think* it should be easy to address though, using the same index/anchor approach used for the string (a bit more code).

File: .github/scripts/update_other_locales.py (hunk 10 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:283]: You have a point that it shouldn't be on this automation to fix the problem.

One issue is that we don't own the code generating the source files:
* iOS: `original source-language target-language datatype`. But Pontoon writes localized files with `original source-language datatype target-language`.
* VPN: `original datatype source-language target-language`. In this case, Pontoon is, for some reason, respecting that order 🤷🏼 

Not doing this would result in a new PR every time the extraction runs, just to change the order of attributes.

The correct fix would be in Pontoon, but it's easier to patch things up on this side. Here's another example: https://github.com/mozilla/pontoon/issues/3517

File: .github/scripts/update_other_locales.py (hunk 14 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py]: Let me take a quick look at how hard it would be to refactor, and what's the performance benefit.

Comment (flodolo) [reply .github/scripts/update_other_locales.py]: So many calls 🤦🏼 

Only ran the script once, for sake of having some data.

Before:
- standard: 1,10s user 0,08s system 85% cpu 1,373 total
- nofile: 2,38s user 0,06s system 99% cpu 2,459 totalù
- matchid: 2,35s user 0,06s system 99% cpu 2,423 total

After:
- standard: 0,82s user 0,05s system 97% cpu 0,887 total
- nofile: 1,78s user 0,06s system 99% cpu 1,853 total
- matchid: 1,72s user 0,06s system 99% cpu 1,805 total

While it's not a massive improvement in practice, it's still a good 30–40%, and definitely worth doing.

File: .github/scripts/update_other_locales.py (hunk 12 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:350]: Missed this comment. Unlikely, but we should be consistent and have a guard here as well.

File: .github/scripts/update_other_locales.py (hunk 9 above)
Comment (flodolo) [reply .github/scripts/update_other_locales.py:392]: I kept this open. I'm not against changing `nofile`, but `matchsource` is not a good replacement. Would `ignorefile`, or `ignore_original` be a better option? Or do we keep `nofile`  knowing the use is limited.

Comment (camilapedraza) [reply .github/scripts/update_other_locales.py:392]: Let's keep nofile. I think we can look into improving this down the road if necessary.

File: .github/scripts/update_other_locales.py (hunk 10 above)
Comment (camilapedraza) [reply .github/scripts/update_other_locales.py:283]: Makes sense. This is something we can look into with the unified moz-l10n lint.

File: .github/scripts/update_other_locales.py (hunk 11 above)
Comment (camilapedraza) [reply .github/scripts/update_other_locales.py:181]: Super clear. Makes sense!

File: .github/scripts/update_other_locales.py [hunk 15]
```diff
@@ -77,9 +77,16 @@ def translation_key(update_type, original_id, source_string):
     return f"{original_id}:{hash(source_string)}"
 
 
-def file_original(trans_node):
-    """Return the 'original' attribute of the <file> a trans-unit belongs to."""
-    return trans_node.xpath("ancestor::x:file", namespaces=NS)[0].get("original")
+def iter_units_by_filenode(root):
```
Comment (camilapedraza) [inline .github/scripts/update_other_locales.py:80]: I rarely see people use yield and it always boggles my mind. Super efficient! 🧑‍🍳💋

**Review by camilapedraza — APPROVED** [review]
This looks ready to ship!
